components/tikv_util/src/trend.rs

```diff
@@ -0,0 +1,737 @@
+//     Include periodic fluctuations, so it's avg could be seen as baseline
+// value     Its duration is also the no-detectable duration after TiKV starting
+// L2:
+//     `L1.avg/L2.avg` to trigger slow-event, last long but low sensitive
+//     Sensitive could be tuned by `L1.duration` and `L2.duration`
+//
+// L* History:
+//     Sample history values and calculate the margin error
+//
+// Spike Filter:
+//     Erase very high and short time spike-values
+//
+pub struct Trend {
+    sample_interval: usize,
+    sample_sequence_id: usize,
+
+    spike_filter: SpikeFilter,
+    spike_filter_enabled: bool,
+
+    data_flow: SampleWindows,
+
+    l1_history: HistoryWindow,
+    l2_history: HistoryWindow,
+
+    // When SPOT is being used, these should be `risk multiple`
+    l1_margin_error_multiple: f64,
+    l2_margin_error_multiple: f64,
+
+    curves_composer: CurvesComposer,
+}
+
+impl Trend {
+    pub fn new(
+        spike_filter_duration: Duration,
+        spike_filter_value_gauge: IntGauge,
+        spike_filter_count_gauge: IntGauge,
+        history_duration: Duration,
+        l0_duration: Duration,
+        l1_duration: Duration,
+        l2_duration: Duration,
+        sample_interval: usize,
+        tolerable_margin_error_value: u64,
+        l1_gap_gauge: IntGauge,
+        l2_gap_gauge: IntGauge,
+        unsensitive_multiple: f64,
+    ) -> Self {
+        let margin_error_base = tolerable_margin_error_value as f64;
+        Self {
+            sample_interval,
+            sample_sequence_id: 0,
+            data_flow: SampleWindows::new(vec![l0_duration, l1_duration, l2_duration]),
+            spike_filter_enabled: !spike_filter_duration.is_zero(),
+            spike_filter: SpikeFilter::new(
+                spike_filter_duration,
+                spike_filter_value_gauge,
+                spike_filter_count_gauge,
+                1.0,
+                5.0,
+                2.0,
+            ),
+            l1_history: HistoryWindow::new(
+                "L1",
+                history_duration,
+                Duration::from_secs(1),
+                margin_error_base,
+                l1_gap_gauge,
+                3.0,
+            ),
+            l2_history: HistoryWindow::new(
+                "L2",
+                history_duration,
+                Duration::from_secs(1),
+                margin_error_base,
+                l2_gap_gauge,
+                2.0,
+            ),
+            l1_margin_error_multiple: 3.0 * unsensitive_multiple,
+            l2_margin_error_multiple: 2.0 * unsensitive_multiple,
+            curves_composer: CurvesComposer::new(l0_duration, l1_duration, l2_duration, 2.0),
+        }
+    }
+
+    #[inline]
+    pub fn record(&mut self, value: u64, now: Instant) {
+        if !self.check_should_sample() {
+            return;
+        }
+        if !self.spike_filter_enabled || !self.data_flow.windows[1].is_overflow() {
```

Comment:
When L1 is full (is_overflow), the window L0 must also be full already (because L0 is smaller):
* At this point the windows L0 and L1 can produce event correctly
* Before L2 is full, the event's duration (produced by L0+L1) will be shorter, but is still correct, so we don't need to wait for L2 is full

So that's why we check and only check L1's fullness here.

Reply:
why don't need to wait L2？

Reply:
The possible cases:
* L2 is full (then L0, L1 are also full) => the event curve of `L0,L1` is valid, curve `L1,L2` is also valid => compose the two curves to one => detecting is funtional 
* L1 is full (then L0 is also full), L2 is not full => the event curve of `L0,L1` is valid, curve `L1,L2` is not valid => use `L0,L1` as event curve => detecting is funtional
* L0 is not full (then L1, L2 are also not full) => no event curve is valid => detecting is **NOT** funtional

So the minimum condition of `detecting is funtional` is `L1 is full`.